### PR TITLE
fix!(cmd_mex_snap): Revert sidegrade functionality and other fixes

### PR DESCRIPTION
**Pull request by badosu**
With the introduction of sidegrade a myriad of regressions were introduced, there were regressions also introduced before that.

This commit reverts it until it can be properly implemented, we also fix the other issues.

Resource spot builder layer is reduced to load before mex snap, we restore the leveraging of its functionalities.

We localize the use of api resources from api_resource_spot_builder and api_resource_spot_finder.

We add a checklist of functionality space for this widget to be checked against:

- [x] **Basic Snap functionality**: Selecting mex as current active
      command shows an arrow towards and a ghost of a mexbuilding to
      snap to closest mex spot

- [x] **Basic Snap functionality**: Clicking places a mex on spot
      ghosted in previous step

- [x] **Avoid already built spots**: If theres a close mex
      spot, but with an already built mex of same tier, it snaps to
      the next closest mex spot

- [x] **Avoid already assigned mex build orders**: If theres a close
      mex spot, but with an order already assigned on the same
      currently selected builder, to build a mex of same tier, it
      snaps to the next closest mex spot.
      _Only when shift is currently pressed (i.e. queued order)_.

- [ ] **Snaps to upgradable mexes**: If theres a close mex
      spot, but with an already built mex of lower tier, it snaps to
      upgrade it to current tier

- [x] **Works on maps with side-loaded mexes**: Basic functionality
      works on maps like Azurite Shores or Rosetta.
      See luarules/gadgets/map_metal_spot_placer.lua for context

### CODE REVIEW

File: luaui/Widgets/cmd_mex_snap.lua [hunk 1]
```diff
@@ -33,36 +70,33 @@ local spGetUnitMetalExtraction = Spring.GetUnitMetalExtraction
...
-				isMexConstructor[uDefID] = uDef.buildOptions
-				break
-			end
-		end
+		unitSizesQuad[uDefID] = { uDef.xsize * 4, uDef.zsize * 4 }
 	end
 end
 
 local function GetExtractionAmount(spot, metalExtracts, orders)
-	local spotWorth = spot.worth/1000.0
```
Comment (badosu) [inline luaui/Widgets/cmd_mex_snap.lua:65]: I assume finder was changed to already do this inside of it, no idea why this was necessary before and not anymore, didn't care to dig what introduced this regression.

File: luaui/Widgets/cmd_mex_snap.lua [hunk 2]
```diff
@@ -369,38 +381,60 @@ function widget:DrawWorld()
...
 	lastMexCmd = os.clock()
 
-	local cbx, cbz = cmdParams[1], cmdParams[3]
-
 	local orders = cmdOpts.shift and GetClashingOrders() or {}
-	local closestSpot = GetClosestMex(bx, bz, WG['resource_spot_finder'].metalSpotsList, metalExtracts, orders)
+	local closestSpot = GetClosestMex(bx, bz, metalSpotsList, metalExtracts, orders)
 
-	if closestSpot then --and not WG['resource_spot_finder'].IsMexPositionValid(closestSpot, cbx, cbz) then
+	if closestSpot then --and not IsMexPositionValid(closestSpot, cmdParams[1], cmdParams[3]) then
```
Comment (badosu) [inline luaui/Widgets/cmd_mex_snap.lua:431]: I kept it commented, but original had this check which was commented in the regression commit, didn't care enough to dig the edge case I used it to address.

**Comment by thehobojoe** [comment]
This code fails at both of these tests:
> Avoid already built spots
> Avoid already assigned mex build orders
spring_EVAYSL9kiW

**Comment by badosu** [comment]
@thehobojoe Can you please provide map, and mex position?

**Comment by badosu** [comment]
@thehobojoe I reproduced and addressed the issue on QuickSilver, please give it a go again. There were small differences between spot finder calculated metal and actual metal from mexes.

Currently it does not snap to upg mexes exactly, it still finds other sides of the spot to build them into.

**Comment by thehobojoe** [comment]
> Currently it does not snap to upg mexes exactly, it still finds other sides of the spot to build them into.

This is intentional right? To remove regressions?

With leaves me with this list:
- Shows arrow to nearest mex, with a ghost at the closest valid location to cursor ✔️ 
- Places mex blueprint at location of ghost on click ✔️ 
- Avoid already built spots ✔️ 
- Avoid already assigned mex build orders ❌ 
  - This worked on quicksilver, but does not work on rosetta. I'm unsure why.
- Works on maps with side-loaded mexes - mostly, only failure I got was the above case.

**Review by thehobojoe — APPROVED** [review]
